Datumizes out jobconfig.toml (extensibility edition)

san7890 opened the pull request:
## About The Pull Request

Hey there,

#75992 (fc54fd6a6079c7c80fd1414ab26bd2e34d0f28bd) made me realize just how wacky the job config system is to people trying to add more new stuff to it, so I finally got the motivation to fix it up.

The gist is that it does all of the same stuff on the front-end as you would expect, but instead of having to modify core generation code to create/recreate the files, it instead uses a robust series of getters and setters. It's much better to use these getters and setters because we can apply needed game logic (like ensuring that the age that the server operator puts in is actually sane, and not completely out of bounds (if you want to permanently price people out of a position, just set it to 0)). The getters are also nifty too, because they let me atone for an early mistake I made with how assistants are meant to work with "unlimited" nonsense.

All a new coder who wants to add stuff to the config needs to do now is create the define, create the datum, add the procs for the datums, and that's it! They don't have to do any of the wacky stuff or account for weird stuff or do any weird copypasta, it's all handled by the system. One datum is all you need, quite neat.

This also fixes some issues in #75992 that probably weren't discovered in testing, but the only good way to fix it was a retune+refactor, which is included in this PR.
## Why It's Good For The Game

Much more extensible, ensuring this system actually works as fully intended, etc.

I fixed up the documentation (they didn't regenerate the job config after they updated the code-side documentation) and some weird spacing stuff that I missed in my review of that aforementioned PR. Everything should work as expected, it's been tested quite a bit. It's also in its own folder now, which is neat because we can share the local defines and split all this stuff out of the already-quite-large SSjob file.
## Changelog
:cl:
server: job_config.toml should now comply with reload-configuration verb, meaning you can hot-reload the configuration from disk and have it apply ingame automatically.
config: The documentation for setting Minimum Character Age on a per-Job basis has been altered to be more explicit.
fix: The Minimum Character Age configuration entry is now sanitized to ensure that it's within the codebase-defined ages, since there's no (legitimate) way to get a character outside of those ages anyways. Invalid values will log to the config log.
/:cl:

san7890 committed:
what the actual glong

san7890 committed:
revertable commit in case i'm being an idiot

san7890 committed:
glurg

san7890 committed:
Revert "revertable commit in case i'm being an idiot"

This reverts commit beaf85175a94ccf3184a94396c79062d4978971e.

san7890 committed:
as anything

san7890 committed:
initial()

san7890 committed:
groundwork for initializing the datums

whoops i gorgot

san7890 committed:
the doohicker

san7890 committed:
full implementation

san7890 committed:
fixes the TOML

san7890 committed:
does some set value stuff for character age

san7890 committed:
proc shuffling

san7890 committed:
last touches

san7890 committed:
eh this list doesn't actually need to be a define

san7890 committed:
I HATE IT HERE

san7890 commented on code/controllers/subsystem/job.dm:
```diff
@@ -605,221 +613,6 @@ SUBSYSTEM_DEF(job)
-
-			if(default_positions || default_positions == 0) // We need to account for jobs that were intentionally turned off via config too.
-				occupation.total_positions = default_positions
-			if(starting_positions || starting_positions == 0)
```
these were replaced with `isnum()` let me know if that's a fuckup, but it seems right to me

LemonInTheDark replied:
it doesn't technically cover as much, but I hope to god none was putting strings in here

san7890 replied:
this comment was before i switched it so that validation of values were done on the datum just in case someone wanted to make string-type config (which was the whole reason why I migrated it from the txt/regex format back in September 2022)

san7890 committed:
file reorganization

* promotes the keys to global defines
* adds the job config datums to a folder in datums
* moves the SSjob work to the configuration folder. i'm okay with this because nothing in this file would be ran without SSconfig. it's all config entry oriented.

san7890 committed:
updates the error string to be a bit more verbose as to what exactly is happening

san7890 committed:
moves the isnum check to the datums

i just remembered i wanted this system to support strings and that sorta thing and that really puts a dampen on that so let's fix that

san7890 committed:
ack i did a half measure

san7890 committed:
compile damn you

LemonInTheDark commented on code/controllers/configuration/entries/jobs.dm:
```diff
@@ -0,0 +1,184 @@
+// This file pretty much just handles all of the interactions between jobconfig.toml and the codebase. This is started by work originating in SSconfig, so I'm okay with it being here.
+
+/// Initializes all of the config singletons for each job config type and adds it to the `job_config_datum_singletons` var list.
+/datum/controller/subsystem/job/proc/generate_config_singletons()
```
s a bit odd to have subsystem procs outside of the subsystem's file, what was the reason for that? just the interaction between this and configs?

san7890 replied:
yes, SSjob was like a thousand lines long of just miscellaneous stuff, felt like it had to do something.

LemonInTheDark commented on code/controllers/configuration/entries/jobs.dm:
```diff
@@ -0,0 +1,184 @@
+/datum/controller/subsystem/job/proc/generate_config_singletons()
+	for(var/datum/job_config_type/config_datum as anything in subtypesof(/datum/job_config_type))
+		var/config_datum_key = initial(config_datum.name)
+		job_config_datum_configurables += config_datum_key
```
why does this exist if job_config_datum_singletons is just it but better

LemonInTheDark commented on code/controllers/configuration/entries/jobs.dm:
```diff
@@ -0,0 +1,184 @@
+
+/// Initializes all of the config singletons for each job config type and adds it to the `job_config_datum_singletons` var list.
+/datum/controller/subsystem/job/proc/generate_config_singletons()
+	for(var/datum/job_config_type/config_datum as anything in subtypesof(/datum/job_config_type))
+		var/config_datum_key = initial(config_datum.name)
+		job_config_datum_configurables += config_datum_key
+		var/new_config_datum = new config_datum
+		job_config_datum_singletons[config_datum_key] = new_config_datum
```
if you do remove job_config_datum_configurables it would be nice to make this proc return a list of key -> datums, rather then just invisibly setting the global list

LemonInTheDark replied:
```suggestion
		job_config_datum_singletons[config_datum_key] = new config_datum
```

san7890 committed:
toml_file -> toml_path

more descript

LemonInTheDark commented on code/controllers/configuration/entries/jobs.dm:
```diff
@@ -0,0 +1,184 @@
+		var/new_config_datum = new config_datum
+		job_config_datum_singletons[config_datum_key] = new_config_datum
+
+/// Called in jobs subsystem initialize if LOAD_JOBS_FROM_TXT config flag is set: reads jobconfig.toml (or if in legacy mode, jobs.txt) to set all of the datum's values to what the server operator wants.
+/datum/controller/subsystem/job/proc/load_jobs_from_config()
```
Describing how something is called is often less interesting then what it does (and is further prone to drift). why do it here?

LemonInTheDark commented on code/controllers/configuration/entries/jobs.dm:
```diff
@@ -0,0 +1,184 @@
+			var/regex/parser = new("[occupation.title]=(-1|\\d+),(-1|\\d+)")
+			parser.Find(jobstext)
+			occupation.total_positions = text2num(parser.group[1])
+			occupation.spawn_positions = text2num(parser.group[2])
```
early returns on the legacy mode please

LemonInTheDark commented on code/controllers/configuration/entries/jobs.dm:
```diff
@@ -0,0 +1,184 @@
+	var/toml_file = "[global.config.directory]/jobconfig.toml"
+
+	if(!legacy_mode) // this flag is set during the setup of SSconfig, and all warnings were handled there.
+		var/job_config = rustg_read_toml_file(toml_file)
```
is this a string? a list? it's typed as nothing but you access it like an assoc list down the line

LemonInTheDark commented on code/controllers/configuration/entries/jobs.dm:
```diff
@@ -0,0 +1,184 @@
+	if(!legacy_mode) // this flag is set during the setup of SSconfig, and all warnings were handled there.
+		var/job_config = rustg_read_toml_file(toml_file)
+
+		for(var/datum/job/occupation as anything in joinable_occupations)
+			var/job_title = occupation.title
+			var/job_key = occupation.config_tag
+			if(!job_config[job_key]) // Job isn't listed, skip it.
+				// List both job_title and job_key in case they de-sync over time.
+				message_admins(span_notice("[job_title] (with config key [job_key]) is missing from jobconfig.toml! Using codebase defaults."))
+				continue
```
bit weird to pull out job_title into its own vars despite it only being used once. I realize this isn't today you, but do you mind removing that?

LemonInTheDark commented on code/controllers/configuration/entries/jobs.dm:
```diff
@@ -0,0 +1,184 @@
+			occupation.total_positions = text2num(parser.group[1])
+			occupation.spawn_positions = text2num(parser.group[2])
+
+/// Called from an admin debug verb that generates the jobconfig.toml file and then allows the end user to download it to their machine.
+/// Returns TRUE if a file is successfully generated, FALSE otherwise.
+/datum/controller/subsystem/job/proc/generate_config(mob/user)
```
similar concern about doccing calls

LemonInTheDark commented on code/controllers/configuration/entries/jobs.dm:
```diff
@@ -0,0 +1,184 @@
+
+		return TRUE
+
+/// If we add a new job or more fields to config a job with, quickly spin up a brand new config that inherits all of your old settings, but adds the new job with codebase defaults.
+/// Returns TRUE if a file is successfully generated, FALSE otherwise.
+/datum/controller/subsystem/job/proc/regenerate_job_config(mob/user)
```
better. describes intent rather then semantics

LemonInTheDark commented on code/controllers/configuration/entries/jobs.dm:
```diff
@@ -0,0 +1,184 @@
+			// Playtime Requirements and Required Account Age are new and we want to see it migrated, so we will just pull codebase defaults for them.
+			// Remember, every time we write the TOML from scratch, we want to have it commented out by default to ensure that the server operator is knows that they codebase defaults when they remove the comment.
+			var/list/working_list = list(
+				"# [JOB_CONFIG_TOTAL_POSITIONS]" = default_positions,
+				"# [JOB_CONFIG_SPAWN_POSITIONS]" = starting_positions,
```
is this the comment in string meme

san7890 replied:
yes. this PR would be absolute hell if both were to be implemented at the same time without any assurance of stability

LemonInTheDark commented on code/controllers/configuration/entries/jobs.dm:
```diff
@@ -0,0 +1,184 @@
+			return FALSE
+		return TRUE
+
+	else // Generate the new TOML format, using codebase defaults.
+		to_chat(user, span_notice("Generating new jobconfig.toml, using codebase defaults."))
+		for(var/datum/job/occupation as anything in joinable_occupations)
+			var/job_key = occupation.config_tag
+			// Remember, every time we write the TOML from scratch, we want to have it commented out by default to ensure that the server operator is knows that they override codebase defaults when they remove the comment.
+			// Having comments mean that we allow server operators to defer to codebase standards when they deem acceptable. They must uncomment to override the codebase default.
+			if(is_assistant_job(occupation)) // there's a concession made in jobs.txt that we should just rapidly account for here I KNOW I KNOW.
+				file_data[job_key] = list(
+					"# [JOB_CONFIG_TOTAL_POSITIONS]" = -1,
+					"# [JOB_CONFIG_SPAWN_POSITIONS]" = -1,
+				)
+				file_data[job_key] += generate_job_config_excluding_legacy(occupation)
+				continue
+
+			// Generate new config from codebase defaults.
+			file_data[job_key] = generate_blank_job_config(occupation)
+
+		if(!export_toml(user, file_data))
+			return FALSE
+
+		return TRUE
```
similar early return

LemonInTheDark commented on code/controllers/configuration/entries/jobs.dm:
```diff
@@ -0,0 +1,184 @@
+
+	var/job_config = rustg_read_toml_file(toml_file)
+	for(var/datum/job/occupation as anything in joinable_occupations)
+		var/job_name = occupation.title
```
similar, weird to pull this out here

LemonInTheDark commented on code/controllers/configuration/entries/jobs.dm:
```diff
@@ -0,0 +1,184 @@
+			continue
+
+		// When we regenerate, we want to make sure commented stuff stays commented, but we also want to migrate information that remains uncommented. So, let's make sure we keep that pattern.
+		if(job_config[job_key]) // Let's see if any data for this job exists.
+			var/list/working_list = list()
+			for(var/config_datum_key in job_config_datum_configurables)
+				var/datum/job_config_type/config_datum = job_config_datum_singletons[config_datum_key]
+				var/config_read_value = job_config[job_key][config_datum_key]
+				if(!config_datum.validate_value(config_read_value))
+					working_list += list(
+						"# [config_datum_key]" = config_datum.get_compile_time_value(occupation), // note that this doesn't make a real comment, it just creates a string mismatch.
+					)
+				else
+					working_list += list(
+						"[config_datum_key]" = config_read_value,
+					)
+
+			file_data[job_key] = working_list
+
+		else
+
+			to_chat(user, span_notice("New job [job_name] (using key [job_key]) detected! Adding to jobconfig.toml using default codebase values..."))
+			file_data[job_key] = generate_blank_job_config(occupation)
```
early continue

LemonInTheDark commented on code/controllers/configuration/entries/jobs.dm:
```diff
@@ -0,0 +1,184 @@
+				var/config_read_value = job_config[job_key][config_datum_key]
+				if(!config_datum.validate_value(config_read_value))
+					working_list += list(
+						"# [config_datum_key]" = config_datum.get_compile_time_value(occupation), // note that this doesn't make a real comment, it just creates a string mismatch.
```
I am depression destroyer of worlds

LemonInTheDark commented on code/controllers/subsystem/job.dm:
```diff
@@ -64,6 +64,12 @@ SUBSYSTEM_DEF(job)
 	/// Dictionary that maps job priorities to low/medium/high. Keys have to be number-strings as assoc lists cannot be indexed by integers. Set in setup_job_lists.
 	var/list/job_priorities_to_strings
 
+	/// List of job config datum singletons.
+	var/static/list/job_config_datum_singletons = list()
+
+	/// List of the types of things that we can configure on a job (practically, the key for the singletons).
+	var/static/list/job_config_datum_configurables = list()
```
as said before, these should be one rather then two
also why are they static? to persist through subsystem rebuild?

san7890 replied:
yeah this is mirroring what SSvote does with its vote singletons since we don't need to waste time newing the datums again

LemonInTheDark replied:
why not just copy them over on subsystem restore?

san7890 replied:
i looked at SSvote again and noticed that it didn't actually use static lists and that i was hallucinating that so i'll just make 'em normal

LemonInTheDark commented on code/datums/job_configs/job_configs.dm:
```diff
@@ -0,0 +1,131 @@
+	CRASH("Attempted to set value for the default job config for [occupation.title] (with config tag [occupation.config_tag])! This is not allowed!")
+
+/// The number of positions a job can have at any given time.
+/datum/job_config_type/default_positions
```
bit weird to have these all in one file, specially when they get their own folder it looks like

LemonInTheDark commented on code/datums/job_configs/job_configs.dm:
```diff
@@ -0,0 +1,131 @@
+		return TRUE
+	return FALSE
+
+/datum/job_config_type/default_positions/set_current_value(datum/job/occupation, value)
+	if(!validate_value(value))
+		return FALSE
+	occupation.total_positions = value
+	return TRUE
```
I don't like that all instances of this proc look the exact same.
could validate_value be called on the parent? 

is there a better way of doing this?

san7890 replied:
i fixed it in an upcoming commit and it's effectively the same boilerplate but it's consistent enough of a pattern where i'm okay with it

LemonInTheDark commented on code/datums/job_configs/job_configs.dm:
```diff
@@ -0,0 +1,131 @@
+	name = JOB_CONFIG_REQUIRED_ACCOUNT_AGE
+
+/datum/job_config_type/required_account_age/get_compile_time_value(datum/job/occupation)
+	return initial(occupation.minimal_player_age)
```
do you think it would be useful to store the linked var name here, and use it automatically (could validate on New())? maybe as a subtype?
A lot of this code looks almost the exact same, which seems inefficient

san7890 replied:
i have no idea what you're talking about here

san7890 replied:
oh i think i know what you mean now, `NAMEOF()` stuff?

san7890 replied:
nope never mind no clue

LemonInTheDark commented on code/datums/job_configs/job_configs.dm:
```diff
@@ -0,0 +1,131 @@
+/datum/job_config_type/required_character_age/get_compile_time_value(datum/job/occupation)
+	return initial(occupation.required_character_age) || 0 // edge case here, this is typically null by default and returning null causes issues. Returning 0 is a safe default.
+
+/datum/job_config_type/required_character_age/validate_value(value) // here we don't care about the value being out of bounds, just wanna make sure it's a number
```
do you want the comment here and not the next line down? having it after the proc def is inconsistent with your pattern in this file

san7890 committed:
single line

san7890 committed:
buncha shit

san7890 committed:
legacy_load()

san7890 committed:
splits it out into files, validate_value on parent

san7890 committed:
whoops

san7890 committed:
another job_name removal

san7890 committed:
nukes that comment

san7890 committed:
types `job_config`

san7890 committed:
creates `import_config_from_txt()`

allows for a nicer looking early return

san7890 committed:
just realized the whole `is_assistant_job()` thing is redundant in the main loop now

whoopie time to save some time

san7890 committed:
touchup

san7890 committed:
split it wrong whoops

san7890 committed:
comment fix

san7890 committed:
supports reloading from disk

forgot to do this the first time around

san7890 committed:
oh whoops i inverted the logic

san7890 committed:
went insane with something during testing but realized

let's fix this

san7890 committed:
vars mutation reading time

we don't need `is_assistant_job()` anymore because that's handled by the overflow stuff. neato.

san7890 committed:
unit testang

san7890 committed:
does the same `vars` thing on the setter proc

requires complete override on `required_character_age` but that's okay.

san7890 committed:
may as well post to job debug

LemonInTheDark commented on code/controllers/configuration/entries/jobs.dm:
```diff
@@ -0,0 +1,185 @@
+// This file pretty much just handles all of the interactions between jobconfig.toml and the codebase. This is started by work originating in SSconfig, so I'm okay with it being here.
+
+/// Initializes all of the config singletons for each job config type and adds it to the `job_config_datum_singletons` var list.
+/datum/controller/subsystem/job/proc/generate_config_singletons()
```
the doc here isn't quite right anymore since it was changed to just return a value

LemonInTheDark commented on code/datums/job_configs/_job_configs.dm:
```diff
@@ -0,0 +1,42 @@
+		qdel(test_occupation)
+
+/// Simply gets the value of the config type for a given job. There can be overrides for special instances on subtypes.
+/datum/job_config_type/proc/get_compile_time_value(datum/job/occupation)
+	return occupation.vars[datum_var_name]
```
Since I'm here I'ma nitpick, would get_default_value be better here, since it's not technically compile time anymore

san7890 replied:
yeah i realized that after i changed it to no longer rely on initial

san7890 committed:
`get_current_value`

san7890 committed:
comment update

Jacquerel commented on code/controllers/configuration/entries/jobs.dm:
```diff
@@ -0,0 +1,185 @@
+		var/starting_positions = text2num(parser.group[2])
+
+		// Playtime Requirements and Required Account Age are new and we want to see it migrated, so we will just pull codebase defaults for them.
+		// Remember, every time we write the TOML from scratch, we want to have it commented out by default to ensure that the server operator is knows that they codebase defaults when they remove the comment.
```
```suggestion
		// Remember, every time we write the TOML from scratch, we want to have it commented out by default to ensure that the server operator knows that they are overriding the codebase defaults when they remove the comment.
```
Is this what this was supposed to say?

Jacquerel commented on code/controllers/configuration/entries/jobs.dm:
```diff
@@ -0,0 +1,185 @@
+	for(var/config_datum_key in job_config_datum_singletons)
+		var/datum/job_config_type/config_datum = job_config_datum_singletons[config_datum_key]
+		// Remember, every time we write the TOML from scratch, we want to have it commented out by default.
+		// This is to ensure that the server operator is knows that they override codebase defaults when they remove the comment.
```
```suggestion
		// This is to ensure that the server operator knows that they are overriding codebase defaults when they remove the comment.
```

Jacquerel reviewed (changes requested):
Looks fine to me chief I'm just not sure what the fuck these comments mean

san7890 committed:
Apply suggestions from code review - Jacquerel edition

Co-authored-by: Jacquerel <hnevard@gmail.com>

san7890 commented:
it's all probably from the old system but i see your point
